### ISSUE TITLE
Suppress subscriptions staging message

### ISSUE THREAD

**Issue by PanosSynetos** [Automattic/wc-calypso-bridge#1048]
On new WooExpress sites, the users see that "it looks like this site has moved or is a duplicate site"

![image](https://user-images.githubusercontent.com/2484390/224927307-ac582c5b-c06f-4db3-abb1-7b0b1bb28373.png)

This doesn't make sense as it's a new site just created. We need to dig into this and discuss with Helix to help

Probably this affects a domain change too.

See related P2 comments peapX7-1V3-p2#comment-2200

**Comment by manospsyx** [comment]
@PanosSynetos I suspect that this will be resolved if WooCommerce Payments is _not_ pre-activated as described in this [PT](https://hostedwoo.wordpress.com/2023/02/24/finalize-validate-free-trial-and-performance-plan-extension-bundles/).

**Comment by james-allan** [comment]
So I'm not super familiar with the way Woo Express sites install plugins but looking at this notice I noticed that the site Subscriptions thinks is the live site is empty. 

![image](https://github.com/Automattic/wc-calypso-bridge/assets/8490476/0d2dd345-bc81-4594-907e-9b009cd399ae)

The final sentence should include the live URL eg: 

"WooCommerce Subscriptions considers example.com to be the site’s URL"

Given the URL is apparently empty and based on @manospsyx's comment above about it being "pre-activated" I assume that subscriptions is being activated on the site before the site has it's eventual URL. 

Is that correct? 

When subscriptions functionality is first activated via WC Subscriptions or WC Payments, it records what it assumes is the store's live URL. If the first-install-url is ever different to the `WP_SITEURL` or value returned from [`get_site_url()`](https://developer.wordpress.org/reference/functions/get_site_url/) it will go into staging mode under the assumption that new URL is the staging or dev url. 

Without knowing how this occurs for Woo Express sites here are some ideas from me: 

1. Like Manos mentioned above, if subscriptions can be activated after the site has the final URL, that should resolve this issue.
2. I don't know if this is even possible, but if it needs to be pre-activated, it is possible to delete the `wc_subscriptions_siteurl` option after it gets assigned the URL? Deleting the option will force subscriptions into resetting it and that should then have the correct site URL.
3. Is there something we can do in the Subscriptions code that would resolve this for you? I don't know enough about the Woo Express flow to know what the solution would look like but, if you have any ideas based on my description of the process, I'm all ears. :)

**Comment by manospsyx** [comment]
Hey @james-allan :) 

Left some thoughts [here](https://hostedwoo.wordpress.com/2023/05/26/my-awesome-store-woo-express-experience/#comment-3090), coping those over:

What happens, in a nutshell, is that:

WooCommerce and the entire list of plugins on a new Woo Express site appear to be first activated at the “Simple Site” phase in the WP.com lifecycle.
Then, the site gets transferred to Atomic, its URL changes, and some code runs that copies over old posts and pages (but post IDs are not maintained, if I recall correctly, [@panossynetos](https://hostedwoo.wordpress.com/mentions/panossynetos/))
As a result of this process, a bunch of things break, such as the associations that WooCommerce maintains with critical Woo pages, and a little bit of [repairing work needs to be done](https://href.li/?https://github.com/Automattic/wc-calypso-bridge/blob/bb937be6794160599605f99c2f77f177e33ea9f1/includes/class-wc-calypso-bridge-setup.php#L47).

It’s the same flow that makes WCS think that the site has changed, and therefore triggers this notice. I believe that there are two problems to solve here.

Prevent WCS from getting into “staging” mode after the Atomic transfer. This can be done by creating a one-time operation like [this](https://github.com/Automattic/wc-calypso-bridge/blob/bb937be6794160599605f99c2f77f177e33ea9f1/includes/class-wc-calypso-bridge-setup.php#L214) — these one-time operations run right after the Atomic transfer, and should be completed on the first page request. If I recall correctly, there is an option that WCS uses to identify the site — I’d probably re-write that.
Prevent WCS from getting into “staging” mode when a merchant purchases a domain. To address this problem, we’d need to re-wire (hook into) the mechanism that WCS uses to identify the site it’s running on, and use an identifier that works better on WP.com — such as the Atomic site ID. Currently, I believe that WCS uses the site URL to identify the site, which can trigger this notice whenever the domain changes.

SomewhereWarm can handle (1), but would appreciate your help to figure out what it’s going to take for WCS to use a different identifier on WP.com (perhaps a filter will be necessary in WCS?).

For us to address (1), we’ll need to know which identifier to utilize. @PanosSynetos if my analysis is sound, then we’ll want to use an identifier unique to Atomic (site/blog ID?). Is there such a thing?

**Comment by manospsyx** [comment]
Bumping priority up to High.

**Comment by james-allan** [comment]
> If I recall correctly, there is an option that WCS uses to identify the site — I’d probably re-write that.

Yeah WC Subscriptions saves what we consider to be the live URL under the `wc_subscriptions_siteurl` option. 

> **Note**
> The URL value stored there is split in half and `_[wc_subscriptions_siteurl]_` is inserted to prevent hosts from doing a find and replace on the database when they create the duplicate site. Probably too many details, but if you you're looking at a value like this `http://wcdev_[wc_subscriptions_siteurl]_.local`, and wondering why it looks like that, that's why. 

Given that flow you mentioned @manospsyx, you should be able to do something similar to [this](https://github.com/Automattic/wc-calypso-bridge/blob/bb937be6794160599605f99c2f77f177e33ea9f1/includes/class-wc-calypso-bridge-setup.php#L214) and call `WCS_Staging::set_duplicate_site_url_lock()` after the site has moved to Atomic. That will force Subscriptions to resave it to the new site's URL.

> 2. Prevent WCS from getting into “staging” mode when a merchant purchases a domain.

Ah good call. When a Woo Express user purchases a domain is there a process similar to when it's transferred to Atomic whereby you could update the option once it is moved to the purchased domain?

If not then yeah we can look into alternative methods you could use for determining if the live site has "moved" under legitimate circumstances. 

FWIW all our code for this functionality lives here: https://github.com/Automattic/woocommerce-subscriptions-core/blob/5.7.1/includes/class-wcs-staging.php

`WCS_Staging::get_live_site_url()` has a filter you could use to override what Subscriptions considers to the live URL:

```php
apply_filters( 'wc_subscriptions_site_url', $url, $path, $scheme, $blog_id ); 
```

The value returned has to be a URL though otherwise our other functionality would break.

**Comment by PanosSynetos** [comment]
OK, we've figured this one out.

It seems like the `upgrade` function of WC Subscriptions, is not running on Atomic sites and is not setting the lock

either this 👇 

```php

// Run upgrades as soon as admin hits site
add_action( 'wp_loaded', __CLASS__ . '::upgrade', 11 );
```

or this 👇 

```php

		// Keep track of site url to prevent duplicate payments from staging sites, first added in 1.3.8 & updated with 1.4.2 to work with WP Engine staging sites
		if ( '0' == self::$active_version || version_compare( self::$active_version, '1.4.2', '<' ) ) {
			WCS_Staging::set_duplicate_site_url_lock();
		}
```
This results in the option `wc_subscriptions_siteurl` being empty

```
woo-boldly-zany-defendor.wpcomstaging.com [150058312] $ wp option get wc_subscriptions_siteurl
Error: Could not get 'wc_subscriptions_siteurl' option. Does it exist?
```

If it doesn't exist, it's considered different by 👇 

```php
			} elseif ( self::get_duplicate_site_lock_key() !== get_option( 'wcs_ignore_duplicate_siteurl_notice' ) ) {
				$notice = new WCS_Admin_Notice( 'error' );
				$notice->set_simple_content(
					sprintf(
						esc_html__( 'It looks like this site has moved or is a duplicate site. .........
```

and we get the notice for new sites.

I'm almost ready with a one time job for this, as @james-allan suggested, and it will run only when it's a new site (5 minutes old) and this `wc_subscriptions_siteurl` option does not exist.

I have an idea for the domain change too, hopefully I'll have something ready for review by tomorrow EOD

**Comment by manospsyx** [comment]
> I have an idea for the domain change too, hopefully I'll have something ready for review by tomorrow EOD

The idea in a nutshell: When the domain changes from `*wpstaging.com` to something else without that string in it, we'll know that a domain upgrade took place, and can swap the site lock key.

@PanosSynetos we could perhaps hook a callback when the `wcs_ignore_duplicate_siteurl_notice` option value is retrieved and filtered to control what's returned, and perhaps even patch the lock key again right at the point when WCS detects a potential mismatch?

**Comment by PanosSynetos** [comment]
Closing the circle - I’ll post in the P2 above and on the issue

I’ve just released WPCOMSH 3.12.19, which includes Calypso Bridge 2.1.8, which includes 😅 

- #1172 

This takes care of the staging message on new sites AND when a domain is purchased.

It will be deployed in about 3-4h from now.
